### PR TITLE
fix: determine type of combinedKeywords

### LINKED ISSUE #233: Object type ignored when nested properties use oneOf and raw
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.23.2

### Plugin version

4.2.1

### Node.js version

20.x

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

13.4

### Description

When I'm creating a schema like this:

```javascript
const schema = S.object().prop(
  "content",
  S.object()
    .raw({ type: "object", discriminator: { propertyName: "type" } })
    .oneOf([schema1, schema2]),
);
```

The `type: object` is omitted in the resulting JSON Schema both event if I use S.object() and raw type object. This can throw errors for example when using Ajv in strict mode. I do not know if the problem it's caused by the oneOf or the raw clause.

### Steps to Reproduce

You can see a full reproduction example here: https://codesandbox.io/p/sandbox/infallible-cloud-4kw5xm. Run npm start.

### Expected Behavior

The resulting JSON schema should not be this:
```
{
  '$schema': 'http://json-schema.org/draft-07/schema#',
  type: 'object',
  properties: { content: { discriminator: [Object], oneOf: [Array] } }
}
```

but this:

```
{
  '$schema': 'http://json-schema.org/draft-07/schema#',
  type: 'object',
  properties: { content: { type: 'object', discriminator: [Object], oneOf: [Array] } }
}
```

It seems that this issue is not present on schemas root level oneOf/raw.

### PR DESCRIPTION
**Pull request by Uzlopak**
Resolves #233 

#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: src/FluentSchema.integration.test.js [hunk 1]
```diff
@@ -224,7 +224,9 @@ describe('S', () => {
   })
 
   describe('compose keywords', () => {
-    const ajv = new Ajv()
+    const ajv = new Ajv({
+      allowUnionTypes: true
```
Comment (mcollina) [inline src/FluentSchema.integration.test.js:228]: Why this change?

File: src/FluentSchema.test.js [hunk 2]
```diff
@@ -202,13 +202,24 @@ describe('S', () => {
...
+    it('anyOf', () => {
+      const schema = S.object()
+        .prop('foo', S.string().anyOf([S.string()]))
+        .valueOf()
+      expect(schema).toEqual({
+        $schema: 'http://json-schema.org/draft-07/schema#',
+        properties: { foo: { type: 'string', anyOf: [{ type: 'string' }] } },
+        type: 'object'
+      })
+    })
```
Comment (mcollina) [inline src/FluentSchema.test.js:214]: Can you add a similar test for oneOf?

**Comment by mcollina** [comment]
Looking at the change, this seems counterintuitive. Why specifying the same thing twice?

File: src/FluentSchema.integration.test.js (hunk 1 above)
Comment (Uzlopak) [reply src/FluentSchema.integration.test.js:228]: Because ajv will warn and spam the log output, because here type will be integer and string.

https://ajv.js.org/strict-mode.html#union-types

**Comment by Uzlopak** [comment]
I just implemented it to fix the reported issue.

**Review by ghost — APPROVED** [review]

**Comment by mcollina** [comment]
Looking at it, it doesn't seems something we should fix. I've actually never used type with anyOf.

**Comment by giovanni-bertoncelli** [comment]
I've not found any JSONSchema specs that actually requires "type" with oneOf.. maybe we should consider raising #233 on AJV side?

**Comment by Uzlopak** [comment]
Closing due to inactivity.

**Comment by giovanni-bertoncelli** [comment]
> I've not found any JSONSchema specs that actually requires "type" with oneOf/anyOf.. maybe we should consider raising #233 on AJV side?

I've been waiting for some reply on this.. I have still a lot of spam in my logs for this issue..

**Comment by climba03003** [comment]
> I've been waiting for some reply on this.. I have still a lot of spam in my logs for this issue..

https://json-schema.org/understanding-json-schema/reference/combining#factoringschemas

`type` is not a requirement when using `oneOf`, `anyOf` or `allOf`.
All the additional properties when using `oneOf`, `anyOf` or `allOf` should be the common part. I don't think it is good to have union type there.
